### PR TITLE
View the details of each bid | Adniya

### PR DESCRIPTION
**Pull request by adniyaYousaf**
User Story 5.3.3

### CODE REVIEW

File: client/src/BidDetail.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+const BidDetail = () => {
+	const [bids, setBids] = useState([]);
+	const [errorMsg, setErrorMsg] = useState(null);
+	const [loading, setLoading] = useState(true);
+	const { bidId } = useParams();
+
+	const fetchBids = async () => {
+		setLoading(true);
+		try {
+			const data = await get("/api/bidder-bid");
```
Comment (soonick) [inline client/src/BidDetail.js]: This is not the right endpoint. You should create an endpoint `/bid/:id`

File: client/src/BidDetail.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
+import { useState, useEffect } from "react";
+import { useParams } from "react-router-dom";
+import { get, post } from "./TenderClient";
+
+const BidDetail = () => {
+	const [bids, setBids] = useState([]);
```
Comment (soonick) [inline client/src/BidDetail.js]:
```suggestion
	const [bid, setBid] = useState({});
```

File: client/src/BidDetail.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	const { bidId } = useParams();
+
+	const fetchBids = async () => {
+		setLoading(true);
+		try {
+			const data = await get("/api/bidder-bid");
+			setBids(data.results);
+			setErrorMsg(null);
+		} catch (error) {
+			setErrorMsg("Error fetching tenders");
```
Comment (soonick) [inline client/src/BidDetail.js]:
```suggestion
			setErrorMsg("Error fetching bid");
```

File: client/src/BidDetail.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		return <div>Loading!!</div>;
+	}
+
+	const results = bids.filter((bid) => bid.bid_id == bidId);
+	const bid = results[0];
+	return (
+		<main>
+			<h2 className="msg">Bid Detail</h2>
+			<div className="container">
+				{errorMsg && <p className="error-message">{errorMsg}</p>}
```
Comment (soonick) [inline client/src/BidDetail.js]: `errorMsg` will never be present here

File: client/src/BidDetail.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+					{bid.status !== "Withdrawn" && (
+						<button
+							className="btn withdraw-btn"
+							onClick={() => handleStatusChange(bid.bid_id, "Withdrawn")}
+						>
+							Withdraw
+						</button>
+					)}
+				</div>
+				{loading && <p>Loading...</p>}
```
Comment (soonick) [inline client/src/BidDetail.js]: `loading` will never be true here

File: client/src/BidDetail.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+					</div>
+
+					{bid.status !== "Withdrawn" && (
+						<button
+							className="btn withdraw-btn"
+							onClick={() => handleStatusChange(bid.bid_id, "Withdrawn")}
+						>
+							Withdraw
+						</button>
+					)}
```
Comment (soonick) [inline client/src/BidDetail.js]: You have the exact same code in two places, move it to a component

**Review by soonick — CHANGES_REQUESTED** [review]

File: client/src/BidDetail.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+const BidDetail = () => {
+	const [bid, setBid] = useState({});
+	const [errorMsg, setErrorMsg] = useState(null);
+	const [loading, setLoading] = useState(true);
+	const { bidId } = useParams();
+
+	const fetchBid = async () => {
+		setLoading(true);
+		try {
+			const data = await get(`/bid/:${bidId}`);
```
Comment (soonick) [inline client/src/BidDetail.js]: Did you test your code?

File: client/src/WithdrawBid.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+const WithdrawBid = ({ bid, handleStatusChange }) => {
```
Comment (soonick) [inline client/src/WithdrawBid.js]: The `handleStatusChange` should also be inside the component.

**Review by soonick — CHANGES_REQUESTED** [review]
We are too close to the deadline. Please remember to test all your code before sending it for review.

File: client/src/BidDetail.js (hunk 7 above)
Comment (adniyaYousaf) [reply client/src/BidDetail.js]: No, but I tested it with `/bidder-bid`. I am working on this endpoint.

Comment (soonick) [reply client/src/BidDetail.js]: Please, don't update PRs until all comments have been addressed, and the code has been tested. Every time you update a PR. I get a notification and review the code. If it's not ready, then it's just a waste of my time

File: client/src/WithdrawBid.js (hunk 8 above)
Comment (adniyaYousaf) [reply client/src/WithdrawBid.js]: I have removed withdraw button because its not in the requirments.

File: client/src/BidDetail.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	};
+
+	useEffect(
+		() => {
+			fetchBid();
+		},
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[bidId]
+	);
```
Comment (soonick) [inline client/src/BidDetail.js]:
```suggestion
	useEffect(
		() => {
			fetchBid(bidId);
		},
		[bidId]
	);
```

File: server/api.js [hunk 10]
```diff
@@ -816,4 +817,42 @@ router.post("/logout", async (req, res) => {
 	}
 });
 
+router.get("/bid-detail/:id", async (req, res) => {
```
Comment (soonick) [inline server/api.js]: This is more standard for rest endpoints:

```suggestion
router.get("/bid/:id", async (req, res) => {
```

File: server/api.js [hunk 11]
```diff
@@ -816,4 +817,42 @@ router.post("/logout", async (req, res) => {
...
+          b.bid_id = $1
+      `,
+			[bidId]
+		);
+
+		if (result.rows.length === 0) {
+			return res.status(404).json({ code: "BID_NOT_FOUND" });
+		}
+
+		res.status(200).json(result.rows[0]);
```
Comment (soonick) [inline server/api.js]: Use our standard response form:

```suggestion
		res.status(200).json({ resource: result.rows[0]});
```

Comment (soonick) [inline server/api.js]: Only the owner of the tender and the owner of the bid should be able to see the details for a bid.

**Review by soonick — CHANGES_REQUESTED** [review]

File: server/api.js (hunk 11 above)
Comment (adniyaYousaf) [reply server/api.js]: i have added permissions in alowlist for this endpoint.
`"/bid/:id": "token",`

File: client/src/BidDetail.js (hunk 9 above)
Comment (adniyaYousaf) [reply client/src/BidDetail.js]: I have tried  but still have formate error.

File: server/api.js (hunk 11 above)
Comment (soonick) [reply server/api.js]: That's not enough. Not all bidder should be able to see all bids. They should only be able to see the bid if they own it.

The same for buyers. Only buyers that own the tender should be able to see bids for that tender.

File: client/src/BidDetail.js (hunk 9 above)
Comment (soonick) [reply client/src/BidDetail.js]: What error?

Comment (adniyaYousaf) [reply client/src/BidDetail.js:30]: Screenshot from 2024-08-09 15-27-50

Comment (soonick) [reply client/src/BidDetail.js]: The answer is right there in the error message

Comment (adniyaYousaf) [reply client/src/BidDetail.js]: I fixed it now.

**Review by soonick — APPROVED** [review]

**Pull request merged**